flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp

```diff
@@ -426,6 +426,93 @@ static bool hasPointerType(mlir::Type type) {
   return type.isa<fir::PointerType>();
 }
 
+// This is a NF performance hack. It makes a simple test that the slices of the
+// load, \p ld, and the merge store, \p st, are trivially mutually exclusive.
+static bool mutuallyExclusiveSliceRange(ArrayLoadOp ld, ArrayMergeStoreOp st) {
+  // If the same array_load, then no further testing is warranted.
+  if (ld.getResult() == st.original())
```

Comment:
What happens for an exact match with no loop carried dependencies, such as `A = A + 1`?  Is a temp needed/generated?

Reply:
No copies. (That case was previously handled.)